### ISSUE TITLE
Uncertainty about pmin calculation

### ISSUE THREAD

**Issue by sneumann** [sneumann/xcms#138]
Hi, from @hpbenton I got this question:

> Dear Steffen,
> 
> With the full xcms rewrite we’re doing some recoding and rewrites ourselves and we’re having a look at the centWave algorithm for improvements and further speed ups. I have a couple of questions that I can’t really understand why some of the code is as it is. Essentially this is surrounding the peak widths for finding the ROI’s. Our questions are around the following:
> 
> Hi Paul,
> 
> I think the pmin calculation (= minCentroids in xcms) used in xcms top level function is strange, see lines 345 - 363 in methods-xcmsRaw.R.
> 
> In essense the following steps are used:
> 
> scalerange <- round((peakwidth / mean(diff(object@scantime))) / 2) (formula A)
> 
> minPeakWidth <-  scales[1] <- scalerange
> 
> minCentroids <- max(4,minPeakWidth-2); (formula B)
> 
> To my opinion pmin must be equal to minimum peak width divided on average time interval netween scans:
> 
> pmin <- round((peakwidth[1] / mean(diff(object@scantime)))) (formula C)
> 
> BUT in (formula A) we get value which is TWO TIMES SMALLER!
> 
> At the same time, in (formula B) we see rather strange expression minPeakWidth - 2 which does not make any sense for me.
> 
> I think here must be minPeakWidth * 2 instead, - in such case we get a natural estimation of pmin as in (formula C).

I have just arrived back from travel, so haven't checked the code yet.

Yours,
Steffen

**Comment by jotsetung** [comment]
I agree, the code is a little confusing, but I think that `pmin <- round((peakwidth[1] / mean(diff(object@scantime))))` (formula C) overestimates the number of required centroids for a peak. Assume e.g. that `peakwidth[1]` is `1` and `mean(diff(object@scantime))` is `0.2790076`. Then `pmin` would be `4`. For the minimum peak width of 1 second, this could however never be fulfilled, because we have only about 3 scans per second in the above example (i.e. `1/mean(diff(object@scantime)) = 3.58`).
So, I think the division by 2 should be OK. Alternatively, a `floor` instead of `round` should also work.

I agree however that the `-2` in formula (B) is a little puzzling.
